### PR TITLE
Implementation of simple while loop parsing

### PR DESCRIPTION
**Pull request by XmiliaH**
My version of the while loop handling.
This version creates phis for all variables prior to the loop.

### CODE REVIEW

File: chapter07/src/main/java/com/seaofnodes/simple/Parser.java [hunk 1]
```diff
@@ -134,14 +135,60 @@ private Node parseBlock() {
...
+        Node ifT = new ProjNode(ifNode, 0, "True" ).peephole();
+        ifNode.unkeep();
+        Node ifF = new ProjNode(ifNode, 1, "False").peephole();
+
+        // Parse the true side, which corresponds to loop body
+        ctrl(ifT);              // set ctrl token to ifTrue projection
+        parseStatement();       // Parse loop body
+
+        if( outerScope.nIns() != _scope.nIns() )
+            throw error("Cannot define a new name in a while loop");
```
Comment (cliffclick) [inline chapter07/src/main/java/com/seaofnodes/simple/Parser.java:180]: well, you ought to be able to define one, with scope just in the loop body.
```java
while( x < lim ) {
  int tmp0 = compute();
  int tmp1 = more();
  x = new_x(tmp0,tmp1);
}         
```

Comment (XmiliaH) [reply chapter07/src/main/java/com/seaofnodes/simple/Parser.java:180]: Yes, this is to catch
```c
while(...) int a = 1;
return a;
```

File: chapter07/src/main/java/com/seaofnodes/simple/node/RegionNode.java [hunk 2]
```diff
@@ -71,5 +74,14 @@ private int findDeadInput() {
         if( lhs==null ) return null;
         _idepth = lhs._idepth+1;
         return (_idom=lhs);
-    } 
+    }
+
+    boolean isFinished() {
```
Comment (cliffclick) [inline chapter07/src/main/java/com/seaofnodes/simple/node/RegionNode.java:79]: yeah, I find the `isFinished` ugly also.  For a non-worklist variant SoN (Simple, right now), it suffices to *not call peephole* until we're done.  As soon as we get a worklist, we'll need something like this.  Right now in AA, I use "is_keep" and look for a special use-flag (the KeepNode uses Nodes I want to keep around).  Elsewhere in Simple we use a `null` use to indicate "I am not yet dead!".

Comment (XmiliaH) [reply chapter07/src/main/java/com/seaofnodes/simple/node/RegionNode.java:79]: The problem with not calling peephole is that an `add(phi, 0)` will idealize to the `phi` and then call the peephole. So there still needs to be some way to stop that.

**Pull request closed without merging**